README.md
# Grafana MQTT panel

- [Grafana MQTT panel](#grafana-mqtt-panel)
  - [MQTT panel](#mqtt-panel)
    - [Some use cases](#some-use-cases)
  - [Tested brokers](#tested-brokers)
  - [Setting up the plugin](#setting-up-the-plugin)
    - [Building the plugin](#building-the-plugin)
    - [Developing the plugin](#developing-the-plugin)

## MQTT panel

This plugin communicates with a MQTT broker through websocket to transmit and receive data.

The panels have 2 types of design, text and gauge.
NB! The unit is not displayed in gauge. Preferred unit for gauge is "short"

It's used to get real time data from equipment, but can be used with anything that can send data to an MQTT broker.

### Some use cases

- Python script that monitors the temperature of the CPU
- Javascript (website) that gets the amount of new visitors

## Tested brokers

- Mosquitto 1.6.9 [How to install mosquitto with websocket (replace 1.4.10 with 1.6.9)](https://gist.github.com/smoofit/dafa493aec8d41ea057370dbfde3f3fc)

## Setting up the plugin

### Building the plugin

First, install dependencies:

```BASH
yarn
```

To build the plugin run:

```BASH
yarn build
```

### Developing the plugin

First, install dependencies:

```BASH
yarn
```

To work with the plugin run:

```BASH
yarn dev
```

or

```BASH
yarn watch
```

Watch will watch for any changes and rebuild the files when it sees any changes to the files


CHANGELOG.md
# Grafana mqtt changelog

- [Grafana mqtt changelog](#grafana-mqtt-changelog)
  - [v0.3.4](#v034)
  - [v0.3.3](#v033)
  - [v0.3.2](#v032)
  - [v0.3.1](#v031)
  - [v0.3.0](#v030)
  - [v0.2.0](#v020)
  - [v0.1.0](#v010)
  - [v0.0.0](#v000)

## v0.3.4

- Plugin.json: Add grafana dependency
- Readme: Add getting started section

## v0.3.3

- Fix wrong url in readme (change screenshot with dist/img)
- Change grafana version dependency to 6.x.x

## v0.3.2

- Fix relative link in readme
- Add logos
- Switch to grafana prettier settings
- Move screenshots to src/img

## v0.3.1

- Added dist for publishing the plugin

## v0.3.0

- Fixed gauge only showing data on "short" unit
- Added better readme description
- Added range option for the gauge

## v0.2.0

- Changed versions to 0 instead of 1 (Beta)
- Changed the plugin name from live data to mqtt

## v0.1.0

- Added gauges
- Added more options
- Added connection status
- Added publish

## v0.0.0

- Inital commit


src copy/piechart_ctrl.test.ts
describe('Basic test', () => {
    const text = 'hello';

    it('should work', () => {
        expect(text).toBe('hello');
    });
});


src/main.ts
// TODO: Fix some numbers shifts the yAxis

import { MetricsPanelCtrl } from 'grafana/app/plugins/sdk';
import * as _ from 'lodash';
import kbn from 'grafana/app/core/utils/kbn';
// @ts-ignore
import { connect } from 'mqtt';

import * as Highcharts from 'highcharts';
require('highcharts/highcharts-more')(Highcharts);
require('highcharts/modules/solid-gauge')(Highcharts);

class Main extends MetricsPanelCtrl {
  static templateUrl = 'module.html';
  $rootScope: any;
  unitFormats: any;
  formattedValue: any;
  value: any;
  unit: any;
  data: any;
  mqttClient: any;
  subscribed: any;
  mqttConnection: any;
  format: any;
  gauge: any;

  /** @ngInject */
  constructor($scope: any, $injector: any, $rootScope: any) {
    super($scope, $injector);
    this.$rootScope = $rootScope;
    this.subscribed = [];
    this.value = 'Waiting for response';
    this.formattedValue = 'Waiting for response';
    this.mqttConnection = {
      message: '',
      status: 'Connect',
    };

    this.format = 'locale';
    const PANEL_DEFAULT = {
      decimals: 2,
      format: 'none',
      mqtt: {
        mode: 'Receive',
        login: {
          hostname: '',
          port: 9001,
        },
        topic: 'data',
        publish: {
          value: 'Hello',
        },
        show: true,
      },
      design: {
        type: 'Text',
      },
      gaugeRange: [0, 100],
    };

    _.defaults(this.panel, PANEL_DEFAULT);

    this.events.on('init-edit-mode', this.onInitEditMode.bind(this));
    this.events.on('panel-initialized', this.initialize.bind(this));
    this.events.on('render', this.onRender.bind(this));
  }

  onRender() {
    if (this.panel.design.type === 'Gauge') {
      if (this.gauge) {
        this.gauge.reflow();
      } else {
        this.createGauge();
      }
    }
  }

  initialize() {
    this.updateMQTTClient();
  }

  createGauge() {
    const PANEL_ELT = document.getElementById(`panel-${this.panel.id}`);
    if (PANEL_ELT) {
      const GAUGE_ELT: any = PANEL_ELT.getElementsByClassName('mqtt-value-gauge')[0];
      if (GAUGE_ELT) {
        const HIGHCHARTS: any = Highcharts;

        this.gauge = HIGHCHARTS.chart(
          HIGHCHARTS.merge(
            {
              chart: {
                renderTo: GAUGE_ELT,
                type: 'solidgauge',
                backgroundColor: 'transparent',
              },

              title: null,

              exporting: {
                enabled: false,
              },

              tooltip: {
                enabled: false,
              },

              // the value axis
              yAxis: {
                stops: [
                  [0.1, '#55BF3B'], // green
                  [0.5, '#DDDF0D'], // yellow
                  [0.9, '#DF5353'], // red
                ],
                lineWidth: 0,
                tickWidth: 0,
                minorTickInterval: null,
                tickAmount: 2,
                labels: {
                  y: 16,
                },
              },

              credits: {
                enabled: false,
              },

              pane: {
                center: ['50%', '75%'],
                size: '140%',
                startAngle: -90,
                endAngle: 90,
                background: [
                  {
                    borderWidth: 1,
                    backgroundColor: 'rgba(100, 100, 100, 0.1)',
                    borderColor: 'rgba(130, 130, 130, 0.4)',
                    innerRadius: '60%',
                    outerRadius: '100%',
                    shape: 'arc',
                  },
                ],
              },

              plotOptions: {
                solidgauge: {
                  dataLabels: {
                    y: -30,
                    borderWidth: 0,
                    useHTML: true,
                  },
                },
              },
            },
            {
              yAxis: {
                title: {
                  text: null,
                },
                labels: {
                  enabled: true,
                },
                startOnTick: false,
                endOnTick: false,
              },

              series: [
                {
                  name: null,
                  data: [0],
                  dataLabels: {
                    format: `<div style="text-align:center"><span>{y:.${this.panel.decimals}f}</span><br/></div>`,
                  },
                },
              ],
            }
          )
        );
        this.updateGaugeRange();
      }
    }
  }

  updateGaugeRange() {
    this.gauge.yAxis[0].setExtremes(...this.panel.gaugeRange);
  }

  getVariableValue(variable: string) {
    if (typeof variable === 'string' && variable !== '' && variable.charAt(0) === '$') {
      const DASHBOARD_VARIABLES = this.templateSrv.variables;
      if (DASHBOARD_VARIABLES.length > 0) {
        const FILTERED_VARIABLES = DASHBOARD_VARIABLES.filter((x: any) => x.name === variable.substr(1));
        if (FILTERED_VARIABLES.length > 0) {
          return FILTERED_VARIABLES[0].current.value;
        }
      }
    }
    return variable;
  }

  updateMQTTClient() {
    const PANEL_ELT = document.getElementById(`panel-${this.panel.id}`);

    if (this.mqttClient) {
      this.mqttClient.end();
    }

    const MQTT_LOGIN_DICT = {
      hostname: this.panel.mqtt.login.hostname,
      port: this.panel.mqtt.login.port,
      username: this.panel.mqtt.login.username,
      password: this.panel.mqtt.login.password,
    };

    const SET_CONNECTION_STATUS = (message: any, hostname: any, port: any, topic: any, status: any) => {
      this.mqttConnection.message = `${message} ${hostname}:${port}/${topic}`;
      this.mqttConnection.status = status;

      const MQTT_CONNECTION_ELT: any = PANEL_ELT
        ? PANEL_ELT.getElementsByClassName('mqtt-connection')
          ? PANEL_ELT.getElementsByClassName('mqtt-connection').length > 0
            ? PANEL_ELT.getElementsByClassName('mqtt-connection')[0]
            : null
          : null
        : null;
      if (MQTT_CONNECTION_ELT) {
        MQTT_CONNECTION_ELT.textContent = this.mqttConnection.message;
      }

      const MQTT_CONNECT_ELT: any = PANEL_ELT
        ? PANEL_ELT.getElementsByClassName('connect-button')
          ? PANEL_ELT.getElementsByClassName('connect-button').length > 0
            ? PANEL_ELT.getElementsByClassName('connect-button')[0]
            : null
          : null
        : null;
      if (MQTT_CONNECT_ELT) {
        MQTT_CONNECT_ELT.textContent = this.mqttConnection.status;
      }
    };

    if (![...Object.values(MQTT_LOGIN_DICT), this.panel.mqtt.topic].includes('')) {
      SET_CONNECTION_STATUS(
        'Preparing to connect to',
        MQTT_LOGIN_DICT.hostname,
        MQTT_LOGIN_DICT.port,
        this.panel.mqtt.topic,
        'Connecting'
      );

      this.mqttClient = connect(MQTT_LOGIN_DICT);

      SET_CONNECTION_STATUS(
        'Connecting to',
        this.mqttClient.options.hostname,
        this.mqttClient.options.port,
        this.panel.mqtt.topic,
        'Connecting'
      );

      this.mqttClient.on('connect', () => {
        this.mqttClient.subscribe(this.panel.mqtt.topic, () => {
          const TOPICS: any = Object.values(this.mqttClient.messageIdToTopic)[0];
          if (TOPICS && TOPICS.length > 0) {
            this.mqttClient.topic = TOPICS[0];
          }
          SET_CONNECTION_STATUS(
            'Connected to',
            this.mqttClient.options.hostname,
            this.mqttClient.options.port,
            this.mqttClient.topic,
            'Connected'
          );
        });
      });

      this.mqttClient.on('error', (err: any) =>
        SET_CONNECTION_STATUS(
          'Failed connecting to',
          this.mqttClient.options.hostname,
          this.mqttClient.options.port,
          this.mqttClient.topic || this.panel.mqtt.topic,
          'Failed'
        )
      );
      this.mqttClient.stream.on('error', (err: any) =>
        SET_CONNECTION_STATUS(
          'Failed connecting to',
          this.mqttClient.options.hostname,
          this.mqttClient.options.port,
          this.mqttClient.topic || this.panel.mqtt.topic,
          'Failed'
        )
      );

      this.mqttClient.on('message', (topic: any, message: any) => {
        this.data = message.toString();
        this.formattedValue = this.formatValue(this.data);
        [this.value, this.unit] = this.formattedValue.split(' ');

        if (this.panel.design.type === 'Text' && PANEL_ELT) {
          const VALUE_ELTS = PANEL_ELT.getElementsByClassName('mqtt-value');
          if (VALUE_ELTS && VALUE_ELTS.length > 0) {
            VALUE_ELTS[0].textContent = this.formattedValue;
          }
        }

        if (!isNaN(this.data)) {
          if (this.panel.design.type === 'Gauge' && this.gauge && this.gauge.series) {
            this.gauge.series[0].points[0].update(parseFloat(this.data));
          }
        }
      });
    }
  }

  onPublish() {
    if (this.mqttClient && this.mqttClient.connected && this.panel.mqtt.topic !== '') {
      this.mqttClient.publish(this.panel.mqtt.topic, this.panel.mqtt.publish.value);
    }
  }

  onInitEditMode() {
    this.addEditorTab('Options', 'public/plugins/gapit-mqtt-panel/editor.html', 2);
    this.unitFormats = kbn.getUnitFormats();
  }

  setUnitFormat(subItem: any) {
    this.panel.format = subItem.value;
    this.value = this.formatValue(this.data);
    this.render();
  }

  getDecimalsForValue(value: any) {
    value = Number(value);
    if (_.isNumber(this.panel.decimals)) {
      return { decimals: this.panel.decimals, scaledDecimals: null };
    }

    const delta = value / 2;
    let dec = -Math.floor(Math.log(delta) / Math.LN10);

    const magn = Math.pow(10, -dec);
    const norm = delta / magn; // norm is between 1.0 and 10.0
    let size;

    if (norm < 1.5) {
      size = 1;
    } else if (norm < 3) {
      size = 2;
      // special case for 2.5, requires an extra decimal
      if (norm > 2.25) {
        size = 2.5;
        ++dec;
      }
    } else if (norm < 7.5) {
      size = 5;
    } else {
      size = 10;
    }

    size *= magn;

    // reduce starting decimals if not needed
    if (Math.floor(value) === value) {
      dec = 0;
    }

    const result = {
      decimals: 0,
      scaledDecimals: 0,
    };
    result.decimals = Math.max(0, dec);
    result.scaledDecimals = result.decimals - Math.floor(Math.log(size) / Math.LN10) + 2;

    return result;
  }

  formatValue(value: any) {
    const decimalInfo = this.getDecimalsForValue(value);
    const formatFunc = kbn.valueFormats[this.panel.format];
    if (formatFunc && !isNaN(Number(value))) {
      return formatFunc(value, decimalInfo.decimals, decimalInfo.scaledDecimals);
    }
    return value;
  }
}

export { Main, Main as MetricsPanelCtrl };


src copy/rendering.ts
import _ from 'lodash';
import './lib/jquery.flot.pie';
import $ from 'jquery';
//import './lib/jquery.flot';

export default function link(scope: any, elem: any, attrs: any, ctrl: any) {
    let data;
    const panel = ctrl.panel;
    elem = elem.find('.piechart-panel__chart');
    const $tooltip = $('<div id="tooltip">') as any;

    ctrl.events.on('render', () => {
        if (panel.legendType === 'Right side') {
            render(false);
            setTimeout(() => {
                render(true);
            }, 50);
        } else {
            render(true);
        }
    });

    function getLegendHeight(panelHeight: any) {
        if (!ctrl.panel.legend.show || ctrl.panel.legendType === 'Right side' || ctrl.panel.legendType === 'On graph') {
            return 20;
        }

        if ((ctrl.panel.legendType === 'Under graph' && ctrl.panel.legend.percentage) || ctrl.panel.legend.values) {
            const breakPoint = parseInt(ctrl.panel.breakPoint, 10) / 100;
            const total = 23 + 20 * data.length;
            return Math.min(total, Math.floor(panelHeight * breakPoint));
        }

        return 0;
    }

    function formatter(label: any, slice: any) {
        const sliceData = slice.data[0][slice.data[0].length - 1];
        let decimal = 2;
        const start = `<div style="font-size:${ctrl.panel.fontSize};text-align:center;padding:2px;">${label}<br/>`;

        if (ctrl.panel.legend.percentageDecimals) {
            decimal = ctrl.panel.legend.percentageDecimals;
        }
        if (ctrl.panel.legend.values && ctrl.panel.legend.percentage) {
            return start + ctrl.formatValue(sliceData) + '<br/>' + slice.percent.toFixed(decimal) + '%</div>';
        } else if (ctrl.panel.legend.values) {
            return start + ctrl.formatValue(sliceData) + '</div>';
        } else if (ctrl.panel.legend.percentage) {
            return start + slice.percent.toFixed(decimal) + '%</div>';
        } else {
            return start + '</div>';
        }
    }

    function noDataPoints() {
        const html = '<div class="datapoints-warning"><span class="small">No data points</span></div>';
        elem.html(html);
    }

    function addPieChart() {
        const width = elem.width();
        const height = ctrl.height - getLegendHeight(ctrl.height);

        const size = Math.min(width, height);

        const plotCanvas = $('<div></div>');
        const plotCss = {
            margin: 'auto',
            position: 'relative',
            paddingBottom: 20 + 'px',
            height: size + 'px',
        };

        plotCanvas.css(plotCss);

        const backgroundColor = $('body').css('background-color');

        const options = {
            legend: {
                show: false,
            },
            series: {
                pie: {
                    radius: 1,
                    innerRadius: 0,
                    show: true,
                    stroke: {
                        color: backgroundColor,
                        width: parseFloat(ctrl.panel.strokeWidth).toFixed(1),
                    },
                    label: {
                        show: ctrl.panel.legend.show && ctrl.panel.legendType === 'On graph',
                        formatter: formatter,
                    },
                    highlight: {
                        opacity: 0.0,
                    },
                    combine: {
                        threshold: ctrl.panel.combine.threshold,
                        label: ctrl.panel.combine.label,
                    },
                },
            },
            grid: {
                hoverable: true,
                clickable: false,
            },
        };

        if (panel.pieType === 'donut') {
            options.series.pie.innerRadius = 0.5;
        }

        data = ctrl.data;

        for (let i = 0; i < data.length; i++) {
            const series = data[i];

            // if hidden remove points
            if (ctrl.hiddenSeries[series.label]) {
                series.data = {};
            }
        }

        if (panel.legend.sort) {
            if (ctrl.panel.valueName !== panel.legend.sort) {
                panel.legend.sort = ctrl.panel.valueName;
            }
            if (panel.legend.sortDesc === true) {
                data.sort((a: any, b: any) => {
                    return b.legendData - a.legendData;
                });
            } else {
                data.sort((a: any, b: any) => {
                    return a.legendData - b.legendData;
                });
            }
        }

        elem.html(plotCanvas);

        // @ts-ignore
        $.plot(plotCanvas, data, options);
        plotCanvas.bind('plothover', (event: any, pos: any, item: any) => {
            if (!item) {
                $tooltip.detach();
                return;
            }

            let body;
            const percent = parseFloat(item.series.percent).toFixed(2);
            const formatted = ctrl.formatValue(item.series.data[0][1]);

            body = '<div class="piechart-tooltip-small"><div class="piechart-tooltip-time">';
            body += '<div class="piechart-tooltip-value">' + _.escape(item.series.label) + ': ' + formatted;
            body += ' (' + percent + '%)' + '</div>';
            body += '</div></div>';

            $tooltip.html(body).place_tt(pos.pageX + 20, pos.pageY);
        });
    }

    function render(incrementRenderCounter: any) {
        if (!ctrl.data) {
            return;
        }

        data = ctrl.data;

        if (0 === ctrl.data.length) {
            noDataPoints();
        } else {
            addPieChart();
        }

        if (incrementRenderCounter) {
            ctrl.renderingCompleted();
        }
    }
}


src/live_data.ts
import { MetricsPanelCtrl } from 'grafana/app/plugins/sdk';
import * as _ from 'lodash';
import kbn from 'grafana/app/core/utils/kbn';
// @ts-ignore
import { connect } from 'mqtt';

import * as Highcharts from 'highcharts';
require('highcharts/highcharts-more')(Highcharts);
require('highcharts/modules/solid-gauge')(Highcharts);

class LiveData extends MetricsPanelCtrl {
    static templateUrl = 'module.html';
    $rootScope: any;
    unitFormats: any;
    value: any;
    data: any;
    mqttClient: any;
    subscribed: any;
    mqttConnection: any;
    format: any;
    gauge: any;

    /** @ngInject */
    constructor($scope: any, $injector: any, $rootScope: any) {
        super($scope, $injector);
        this.$rootScope = $rootScope;
        this.subscribed = [];
        this.value = 'Waiting for response';
        this.mqttConnection = {
            message: '',
            status: 'Connect'
        };

        this.format = 'locale';
        const PANEL_DEFAULT = {
            mqtt: {
                mode: 'Recieve',
                login: {
                    hostname: '',
                    port: 9001
                },
                topic: 'data',
                publish: {
                    value: 'Hello'
                },
                show: true
            },
            design: {
                type: 'Text'
            }
        };

        _.defaults(this.panel, PANEL_DEFAULT);

        this.events.on('init-edit-mode', this.onInitEditMode.bind(this));
        this.events.on('panel-initialized', this.initialize.bind(this));
        this.events.on('render', this.onRender.bind(this));
    }

    onRender() {
        if (this.panel.design.type == 'Gauge') {
            if (this.gauge) {
                this.gauge.reflow();
            } else {
                this.createGauge();
            }
        }
    }

    initialize() {
        this.updateMQTTClient();
        this.createGauge();
    }

    createGauge() {
        if (!this.gauge) {
            const PANEL_ELT = document.getElementById(`panel-${this.panel.id}`);
            if (PANEL_ELT) {
                const GAUGE_ELT: any = PANEL_ELT.getElementsByClassName('live-data-value-gauge')[0];
                if (GAUGE_ELT) {
                    const HIGHCHARTS: any = Highcharts;

                    this.gauge = HIGHCHARTS.chart(
                        HIGHCHARTS.merge(
                            {
                                chart: {
                                    renderTo: GAUGE_ELT,
                                    type: 'solidgauge',
                                    backgroundColor: 'transparent'
                                },

                                title: null,


                                exporting: {
                                    enabled: false
                                },

                                tooltip: {
                                    enabled: false
                                },

                                // the value axis
                                yAxis: {
                                    stops: [
                                        [0.1, '#55BF3B'], // green
                                        [0.5, '#DDDF0D'], // yellow
                                        [0.9, '#DF5353'] // red
                                    ],
                                    lineWidth: 0,
                                    tickWidth: 0,
                                    minorTickInterval: null,
                                    tickAmount: 2,
                                    labels: {
                                        y: 16
                                    }
                                },

                                credits: {
                                    enabled: false
                                },

                                pane: {
                                    background: [{
                                        borderWidth: 0,
                                        backgroundColor: 'transparent'
                                    }]
                                },

                                plotOptions: {
                                    solidgauge: {
                                        dataLabels: {
                                            y: -20,
                                            borderWidth: 0,
                                            useHTML: true
                                        }
                                    }
                                }
                            },
                            {
                                yAxis: {
                                    min: 0,
                                    max: 50,
                                    title: {
                                        text: null
                                    },
                                    labels: {
                                        enabled: false
                                    }
                                },

                                series: [
                                    {
                                        name: null,
                                        data: [0],
                                        dataLabels: {
                                            format:
                                                '<div style="text-align:center"><span>{y:.1f}</span><br/></div>'
                                        }
                                    }
                                ]
                            }
                        )
                    );
                }
            }
        }
    }

    getVariableValue(variable: string) {
        if (typeof(variable) == 'string' && variable != '' && variable.charAt(0) == '$') {
            const DASHBOARD_VARIABLES = this.templateSrv.variables;
            if (DASHBOARD_VARIABLES.length > 0) {
                const FILTERED_VARIABLES = DASHBOARD_VARIABLES.filter((x: any) => x.name == variable.substr(1))
                if (FILTERED_VARIABLES.length > 0) {
                    return FILTERED_VARIABLES[0].current.value;
                }
            }
        }
        return variable;
    }

    updateMQTTClient() {
        console.log(this.templateSrv.variables.filter((x: any) => x.name == "test")[0].current.value);
        const PANEL_ELT = document.getElementById(`panel-${this.panel.id}`);

        if (this.mqttClient) {
            this.mqttClient.end();
        }

        const MQTT_CONVERTED_DICT = {
            hostname: this.getVariableValue(this.panel.mqtt.login.hostname),
            port: this.getVariableValue(this.panel.mqtt.login.port),
            username: this.getVariableValue(this.panel.mqtt.login.username),
            password: this.getVariableValue(this.panel.mqtt.login.password),
            topic: this.getVariableValue(this.panel.mqtt.topic)
        };
        console.log(MQTT_CONVERTED_DICT);

        const MQTT_LOGIN_DICT = {
            hostname: this.panel.mqtt.login.hostname,
            port: this.panel.mqtt.login.port,
            username: this.panel.mqtt.login.username,
            password: this.panel.mqtt.login.password,
        };

        const SET_CONNECTION_STATUS = (
            message: any,
            hostname: any,
            port: any,
            topic: any,
            status: any
        ) => {
            this.mqttConnection.message = `${message} ${hostname}:${port}/${topic}`;
            this.mqttConnection.status = status;

            const MQTT_CONNECTION_ELT: any = PANEL_ELT
                ? PANEL_ELT.getElementsByClassName('mqtt-connection')
                    ? PANEL_ELT.getElementsByClassName('mqtt-connection')
                        .length > 0
                        ? PANEL_ELT.getElementsByClassName('mqtt-connection')[0]
                        : null
                    : null
                : null;
            if (MQTT_CONNECTION_ELT) {
                MQTT_CONNECTION_ELT.textContent = this.mqttConnection.message;
            }

            const MQTT_CONNECT_ELT: any = PANEL_ELT
                ? PANEL_ELT.getElementsByClassName('connect-button')
                    ? PANEL_ELT.getElementsByClassName('connect-button')
                        .length > 0
                        ? PANEL_ELT.getElementsByClassName('connect-button')[0]
                        : null
                    : null
                : null;
            if (MQTT_CONNECT_ELT) {
                MQTT_CONNECT_ELT.textContent = this.mqttConnection.status;
            }
        };

        if (
            ![
                ...Object.values(MQTT_LOGIN_DICT),
                this.panel.mqtt.topic
            ].includes('')
        ) {
            SET_CONNECTION_STATUS(
                'Preparing to connect to',
                MQTT_LOGIN_DICT.hostname,
                MQTT_LOGIN_DICT.port,
                this.panel.mqtt.topic,
                'Connecting'
            );

            this.mqttClient = connect(MQTT_LOGIN_DICT);

            SET_CONNECTION_STATUS(
                'Connecting to',
                this.mqttClient.options.hostname,
                this.mqttClient.options.port,
                this.panel.mqtt.topic,
                'Connecting'
            );

            this.mqttClient.on('connect', () => {
                this.mqttClient.subscribe(this.panel.mqtt.topic, () => {
                    const TOPICS: any = Object.values(
                        this.mqttClient.messageIdToTopic
                    )[0];
                    if (TOPICS && TOPICS.length > 0) {
                        this.mqttClient.topic = TOPICS[0];
                    }
                    SET_CONNECTION_STATUS(
                        'Connected to',
                        this.mqttClient.options.hostname,
                        this.mqttClient.options.port,
                        this.mqttClient.topic,
                        'Connected'
                    );
                });
            });

            this.mqttClient.on('error', (err: any) =>
                SET_CONNECTION_STATUS(
                    'Failed connecting to',
                    this.mqttClient.options.hostname,
                    this.mqttClient.options.port,
                    this.mqttClient.topic || this.panel.mqtt.topic,
                    'Failed'
                )
            );
            this.mqttClient.stream.on('error', (err: any) =>
                SET_CONNECTION_STATUS(
                    'Failed connecting to',
                    this.mqttClient.options.hostname,
                    this.mqttClient.options.port,
                    this.mqttClient.topic || this.panel.mqtt.topic,
                    'Failed'
                )
            );

            this.mqttClient.on('message', (topic: any, message: any) => {
                this.data = message.toString();
                this.value = this.formatValue(this.data);
                if (this.panel.design.type == 'Text' && PANEL_ELT) {
                    const VALUE_ELTS = PANEL_ELT.getElementsByClassName("live-data-value");
                    if (VALUE_ELTS && VALUE_ELTS.length > 0) {
                        VALUE_ELTS[0].textContent = this.value;
                    }
                }
                if (!isNaN(this.value)) {
                    if (this.panel.design.type == 'Gauge' && this.gauge && this.gauge.series) {
                        this.gauge.series[0].points[0].update(
                            parseFloat(this.value)
                        );
                    }
                }
            });
        }
    }

    onPublish() {
        if (
            this.mqttClient &&
            this.mqttClient.connected &&
            this.panel.mqtt.topic != ''
        ) {
            this.mqttClient.publish(
                this.panel.mqtt.topic,
                this.panel.mqtt.publish.value
            );
        }
    }

    onInitEditMode() {
        this.addEditorTab(
            'Options',
            'public/plugins/gapit-live_data-panel/editor.html',
            2
        );
        this.unitFormats = kbn.getUnitFormats();
    }

    setUnitFormat(subItem: any) {
        this.panel.format = subItem.value;
        this.value = this.formatValue(this.data);
        this.render();
    }

    getDecimalsForValue(value: any) {
        value = Number(value);
        if (_.isNumber(this.panel.decimals)) {
            return { decimals: this.panel.decimals, scaledDecimals: null };
        }

        const delta = value / 2;
        let dec = -Math.floor(Math.log(delta) / Math.LN10);

        const magn = Math.pow(10, -dec);
        const norm = delta / magn; // norm is between 1.0 and 10.0
        let size;

        if (norm < 1.5) {
            size = 1;
        } else if (norm < 3) {
            size = 2;
            // special case for 2.5, requires an extra decimal
            if (norm > 2.25) {
                size = 2.5;
                ++dec;
            }
        } else if (norm < 7.5) {
            size = 5;
        } else {
            size = 10;
        }

        size *= magn;

        // reduce starting decimals if not needed
        if (Math.floor(value) === value) {
            dec = 0;
        }

        const result = {
            decimals: 0,
            scaledDecimals: 0
        };
        result.decimals = Math.max(0, dec);
        result.scaledDecimals =
            result.decimals - Math.floor(Math.log(size) / Math.LN10) + 2;

        return result;
    }

    formatValue(value: any) {
        const decimalInfo = this.getDecimalsForValue(value);
        const formatFunc = kbn.valueFormats[this.panel.format];
        if (formatFunc && !isNaN(Number(value))) {
            return formatFunc(
                value,
                decimalInfo.decimals,
                decimalInfo.scaledDecimals
            );
        }
        return value;
    }
}

export { LiveData, LiveData as MetricsPanelCtrl };
